sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp

```diff
@@ -21,21 +29,140 @@ namespace Azure { namespace Core { namespace Test {
    */
   class TestBase : public ::testing::Test {
 
+  private:
+    // Call this method to update client options with the required configuration to
+    // support Record & Playback.
+    // If Playback or Record is not set, no changes will be done to the clientOptions or credential.
+    // Call this before creating the SDK client
+    void PrepareClientOptions(
+        std::shared_ptr<Core::Credentials::TokenCredential>** credential,
```

Comment:
shared_ptr creates a copy of the ptr every time it is passed as a parameter.  This function wants to update the original value and not the copied shared_ptr

see the next simple example:


void Foo(std::shared_ptr<std::string> prt) {

     //  This function wants to update the ptr so it forgets its current value and now point to another string.
    //  However, it is updating only the copied ptr in the arg, and not the original.
     ptr = std::make_shared<std::string>(" Another value ");
}

void Foo2(std::shared_ptr<std::string> * ptr) {
     // This function will update the original!
     *ptr = std::make_shared<std::string>(" Another value "); 
}

auto originalPtr = std::make_shared<std::string>(" the original value");

//  originalPtr won't be updated, only it's copy
Foo(originalPtr);

// Passing the address (a pointer to the ptr will update the original )
Foo2(&originalPtr);


Lastly, imagine `Foo2` wants to pass the pointer to another function to update the original value ( delegate the change) like:

void Foo2(std::shared_ptr<std::string> * ptr) {
     // Let Foo3 make the change instead of this function,
     Foo3(ptr); 
}

Can you imagine how Foo3 should be?  If Foo3 receives a simple pointer ( * ), it would make a copy of whatever pointer it gets and reassigning that pointer won't work.  That's why a Foo3 would need to receive a pointer to a pointer like:

void Foo3(std::shared_ptr<std::string>** ptr) {
     // Now we can change the original initial ptr by doing
    **ptr = std::make_shared<std::string>(" other value");
}

Reply:
> shared_ptr creates a copy of the ptr every time it is passed as a parameter.

😯 I see

So, if `Foo3` wanted to delegate the change to the original value to another function, `Foo4`, would we pass in `std::shared_ptr<std::string>*** ptr`?

Is there value in using a `shared_ptr` here, instead of just a regular ptr?

@CaseyCarter I am not familiar with this pattern, how does it look to you?

Reply:
What's wrong with `std::shared_ptr<Core::Credentials::TokenCredential>& credential`?